### DISCUSSION TITLE
How to use GPIOs?

### DISCUSSION THREAD

**Discussion by rei-vilo** [phoddie/node-red-mcu#35] [Q&A] [answered]
Now that Node-RED MCU is configured and running (see #26), I would like the blink the built-in LED of the Raspberry Pi Pico W.

How to address a GPIO? Which node should be used?

The [GPIO-out](https://github.com/phoddie/node-red-mcu#gpio-out) section doesn't say much.

* On the Raspberry Pi Pico, built-in LED is `GPIO25`.
* On the Raspberry Pi Pico W, built-in LED is not managed directly, but by the WiFi CYW43439 radio `WL_GPIO0`.

Thank you!

**Comment by phoddie** [comment]
As you note, the Pico W doesn't connect the on-board LED directly to the MCU. Instead, it routes through the Wi-Fi part. To accommodate this hardware design, the [Digital](https://419.ecma-international.org/#-10-io-classes-digital) constructor supports that pin on GPIO bank 1. The LED is pin 0 of bank 1, so use 32 as the pin with the [rpi-gpio out](url) Node ((32 * 1) + 0). (I don't have a board with me to try, so this is based on reviewing the [implementation](https://github.com/Moddable-OpenSource/moddable/blob/3c7f16c3c113aefa907b79a1884ab09a5b8a9ab9/modules/io/digital/pico/digitalbank.c#L331-L347)!)

**Comment by rei-vilo** [comment]
Using pin 32 doesn't seem to work.

<img width="704" alt="Capture 2022-11-02 à 09 40 31" src="https://user-images.githubusercontent.com/951776/199440755-ba6917d8-5455-4670-bcc7-fa659097e7fd.png">

Neither other standard pins, like GPIO15.

**Reply by phoddie** [reply]
Please share the JSON of your flow. Thank you.

**Reply by rei-vilo** [reply]
Sure!

```
[
    {
        "id": "b0430bb22e8d2233",
        "type": "inject",
        "z": "fa3bf03c6582e230",
        "name": "",
        "props": [],
        "repeat": "2",
        "crontab": "",
        "once": true,
        "onceDelay": "1",
        "topic": "",
        "_mcu": {
            "mcu": true
        },
        "x": 130,
        "y": 100,
        "wires": [
            [
                "a3c28be7578a24fc",
                "ea51a1713c293c29"
            ]
        ]
    },
    {
        "id": "a3c28be7578a24fc",
        "type": "delay",
        "z": "fa3bf03c6582e230",
        "name": "",
        "pauseType": "delay",
        "timeout": "500",
        "timeoutUnits": "milliseconds",
        "rate": "1",
        "nbRateUnits": "1",
        "rateUnits": "second",
        "randomFirst": "1",
        "randomLast": "5",
        "randomUnits": "seconds",
        "drop": false,
        "allowrate": false,
        "outputs": 1,
        "_mcu": {
            "mcu": true
        },
        "x": 190,
        "y": 240,
        "wires": [
            [
                "26da7f94dc0f4b7e"
            ]
        ]
    },
    {
        "id": "65c9d70eff75522b",
        "type": "rpi-gpio out",
        "z": "fa3bf03c6582e230",
        "name": "",
        "pin": "32",
        "set": "",
        "level": "0",
        "freq": "",
        "out": "out",
        "bcm": true,
        "_mcu": {
            "mcu": true
        },
        "x": 540,
        "y": 180,
        "wires": []
    },
    {
        "id": "f26c11b6a04cbfd0",
        "type": "rpi-gpio out",
        "z": "fa3bf03c6582e230",
        "name": "",
        "pin": "32",
        "set": "",
        "level": "0",
        "freq": "",
        "out": "out",
        "bcm": true,
        "_mcu": {
            "mcu": true
        },
        "x": 540,
        "y": 280,
        "wires": []
    },
    {
        "id": "ea51a1713c293c29",
        "type": "change",
        "z": "fa3bf03c6582e230",
        "name": "1",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "1",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "_mcu": {
            "mcu": true
        },
        "x": 370,
        "y": 160,
        "wires": [
            [
                "65c9d70eff75522b",
                "9aad4231d9142b20"
            ]
        ]
    },
    {
        "id": "26da7f94dc0f4b7e",
        "type": "change",
        "z": "fa3bf03c6582e230",
        "name": "0",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "0",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "_mcu": {
            "mcu": true
        },
        "x": 370,
        "y": 260,
        "wires": [
            [
                "f26c11b6a04cbfd0",
                "0ae1e6d23a6488ce"
            ]
        ]
    },
    {
        "id": "9aad4231d9142b20",
        "type": "debug",
        "z": "fa3bf03c6582e230",
        "name": "debug 1",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 540,
        "y": 140,
        "wires": []
    },
    {
        "id": "0ae1e6d23a6488ce",
        "type": "debug",
        "z": "fa3bf03c6582e230",
        "name": "debug 2",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 540,
        "y": 240,
        "wires": []
    }
]
```

I also connected an LED to GPIO15 and entered `15`, to no avail.

<img width="598" alt="Capture 2022-11-02 à 16 49 03" src="https://user-images.githubusercontent.com/951776/199536577-c9cbc8a0-69d0-4ba9-80e9-15aa3c657fa8.png">

**Reply by phoddie** [reply]
This was fun. Three issues!

1. The flow isn't quite correct. It uses GPIO 32 in two nodes. The implementation of RPi Digital Out requires that each GPIO be owned by a single node. An exception is thrown when the second node tries to use GPIO 32. That exception is unhandled so execution stops.  Here's a variation of your flow to use a single GPIO (it uses a `junction`, so be sure that you are using Node-RED 3.x).
<details>
<summary>Revised GPIO blink flow</summary>

```json
[
    {
        "id": "62c8b602489dac4b",
        "type": "tab",
        "label": "gpio test",
        "disabled": false,
        "info": "",
        "env": []
    },
    {
        "id": "08c2c914966facb9",
        "type": "junction",
        "z": "62c8b602489dac4b",
        "x": 600,
        "y": 120,
        "wires": [
            [
                "65c9d70eff75522b",
                "9aad4231d9142b20"
            ]
        ]
    },
    {
        "id": "b0430bb22e8d2233",
        "type": "inject",
        "z": "62c8b602489dac4b",
        "name": "",
        "props": [],
        "repeat": "2",
        "crontab": "",
        "once": true,
        "onceDelay": "1",
        "topic": "",
        "_mcu": false,
        "x": 150,
        "y": 80,
        "wires": [
            [
                "a3c28be7578a24fc",
                "ea51a1713c293c29"
            ]
        ]
    },
    {
        "id": "a3c28be7578a24fc",
        "type": "delay",
        "z": "62c8b602489dac4b",
        "name": "",
        "pauseType": "delay",
        "timeout": "500",
        "timeoutUnits": "milliseconds",
        "rate": "1",
        "nbRateUnits": "1",
        "rateUnits": "second",
        "randomFirst": "1",
        "randomLast": "5",
        "randomUnits": "seconds",
        "drop": false,
        "allowrate": false,
        "outputs": 1,
        "_mcu": false,
        "x": 330,
        "y": 160,
        "wires": [
            [
                "26da7f94dc0f4b7e"
            ]
        ]
    },
    {
        "id": "65c9d70eff75522b",
        "type": "rpi-gpio out",
        "z": "62c8b602489dac4b",
        "name": "",
        "pin": "32",
        "set": "",
        "level": "0",
        "freq": "",
        "out": "out",
        "bcm": true,
        "_mcu": false,
        "x": 720,
        "y": 160,
        "wires": []
    },
    {
        "id": "ea51a1713c293c29",
        "type": "change",
        "z": "62c8b602489dac4b",
        "name": "1",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "1",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "_mcu": false,
        "x": 490,
        "y": 80,
        "wires": [
            [
                "08c2c914966facb9"
            ]
        ]
    },
    {
        "id": "26da7f94dc0f4b7e",
        "type": "change",
        "z": "62c8b602489dac4b",
        "name": "0",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "0",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "_mcu": false,
        "x": 490,
        "y": 160,
        "wires": [
            [
                "08c2c914966facb9"
            ]
        ]
    },
    {
        "id": "9aad4231d9142b20",
        "type": "debug",
        "z": "62c8b602489dac4b",
        "name": "debug 1",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": false,
        "x": 720,
        "y": 80,
        "wires": []
    }
]
```

</details>

2. GPIO 32 is special on Raspberry Pi Pico because it is not connected directly to the MCU, but is on the Wi-Fi component. The Wi-Fi subsystem must be initialized to access GPIO 32(!). So, GPIO 32 worked if Wi-Fi was initialized (SSID set) and otherwise would hang. A fix has been committed to the Moddable SDK for that. See the Updating section for your development platform in our [Getting Started](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md) guide.
3. The unhandled exception mentioned in (1) above doesn't seem to be reported by the Node-RED MCU plug-in, so you probably didn't notice it. We should see if @ralphwetzel can update the proxy and plug-in so that information is displayed to the user in the editor. (Even if we eliminate this particularly exception, there will always be some unhandled exceptions so best for the plug-in to propagate that information reliably).

**Reply by rei-vilo** [reply]
Thank you for the answer.

1. I am using Node-RED v3.0.2.
2. I have update the Moddable SDK.
3. I tried the revised GPIO blink flow.

Unfortunately, it raises an error:

> sntp was not defined. call done.
> /Users/USER/.node-red/node_modules/@ralphwetzel/node-red-mcu-plugin/mcu_modules/require/require.js (5) # Break: (host): module "child_process" not found!
> XS abort: unhandled exception

<img width="872" alt="Capture 2022-11-03 à 12 33 09" src="https://user-images.githubusercontent.com/951776/199711062-1b8a335e-57cc-4640-bad4-5324b21cce7c.png">

To be continued...

**Reply by ralphwetzel** [reply]
> Unfortunately, it raises an error:

I was already impressed how both of you were able to build this flow - as the plugin creates an erroneous manifest  for `rpi-gpio out` (despite there's no need at all to create one!  😏 ). This is the error you hit here. A fix is on it's way...

**Reply by phoddie** [reply]
Thanks, @ralphwetzel. I had no idea where the "child_process" module was coming from. That's not a Moddable SDK or Node-RED MCU Edition module!

**Reply by ralphwetzel** [reply]
@phoddie --
> We should see if @ralphwetzel can update the proxy and plug-in so that information is displayed to the user in the editor.

I thought about intercepting the message that the mcu sends to xsbug. Unfortunately this gives us only the error message - yet not the location of the error (== node id!). The best that can be done in this case is to display the message in the `Console Monitor`.

To get a behavior that mimics standard Node-RED demands to catch the error in the node, wrap "it" (error.message should be enough) in a trace container and send this to the editor:

```javascript
    // Example: CompatibilityNode
    onMessage(msg, done) {
        try { 
            this.#events.input?.forEach(input => input.call(this, msg, this.#send, done));
        } catch(err) {
            trace.left(JSON.stringify({
                "error": {
                    "error": err.toString(),
                }
            }), node.id);
            // There should be a log / error message for xsbug as well...
        }
    }
```
This will create a standard error indication in Node-RED's Debug pane.

**Reply by phoddie** [reply]
Just FYI - I confirmed that the "rpi gpio" plugin allows several nodes to share a single GPIO, as long as they are all of the same type (input, output, PWM, etc). The MCU implementation now does the same, so that the flow @rei-vilo shared above (with two GPIO outputs connected to the same GPIO) now works without modification.

**Comment by rei-vilo** [comment, marked as answer]
Thank you so much for your help. 

I can now turn the LED of the Raspberry Pi Pico W on and off through MQTT.

To be continued...
